web/src/main/java/com/bazaarvoice/emodb/web/throttling/DataStoreUpdateThrottleControlTask.java

```diff
@@ -0,0 +1,154 @@
+ * use the wildcard API key, "*".
+ *
+ * Usage:
+ * <pre>
+ *     # Set a throttle for a single API key, with the default and with an explicit expiration
+ *     curl -s -XPOST 'localhost:8081/tasks/sor-api-update-throttle?id=ID01&limit=100'
+ *     curl -s -XPOST 'localhost:8081/tasks/sor-api-update-throttle?id=ID02&limit=100&duration=PT1H'
+ *
+ *     # Set an instance wide throttle
+ *     curl -s -XPOST 'localhost:8081/tasks/sor-api-update-throttle?id=*&limit=500&duration=PT30M'
+ *
+ *     # Remove a throttle
+ *      curl -s -XPOST 'localhost:8081/tasks/sor-api-update-throttle?id=ID02&limit=0'
+ *
+ *     # List all throttled endpoints
+ *      curl -s -XPOST 'localhost:8081/tasks/sor-api-update-throttle'
+ *
+ *     # Clear all throttled endpoints
+ *     curl -s -XPOST 'localhost:8081/tasks/sor-api-update-throttle?clear'
+ * </pre>
+ */
+public class DataStoreUpdateThrottleControlTask extends Task {
+    private final DataStoreUpdateThrottleManager _throttleManager;
+
+    @Inject
+    public DataStoreUpdateThrottleControlTask(TaskRegistry taskRegistry,
+                                              DataStoreUpdateThrottleManager throttleManager) {
+        super("sor-api-update-throttle");
+        _throttleManager = requireNonNull(throttleManager, "throttleManager");
+        taskRegistry.addTask(this);
+    }
+
+    @Override
+    public void execute(ImmutableMultimap<String, String> parameters, PrintWriter out) throws Exception {
+        // Default duration
+        Duration expiryDuration = Duration.ofHours(24);
+        Collection<String> values = parameters.get("duration");
+        if (values.size() == 1) {
+            expiryDuration = new ZkDurationSerializer().fromString(values.iterator().next());
+        } else if (values.size() > 1) {
+            throw new IllegalArgumentException("At most one duration parameter is permitted");
+        }
+
+        double limit = 0;
+        values = parameters.get("limit");
+        if (values.size() == 1) {
+            limit = Double.parseDouble(values.iterator().next());
+        } else if (values.size() > 1) {
+            throw new IllegalArgumentException("At most one limit parameter is permitted");
+        }
+
+        boolean throttlesChanged = false;
+
+        if (parameters.keySet().contains("clear")) {
+            for (String apiKey : _throttleManager.getAllThrottles().keySet()) {
+                _throttleManager.clearAPIKeyRateLimit(apiKey);
+            }
+            out.println("All throttles cleared");
+            throttlesChanged = true;
+        }
+
+        if (!parameters.get("id").isEmpty()) {
+            DataStoreUpdateThrottle throttle = null;
+            if (limit > 0) {
+                throttle = new DataStoreUpdateThrottle(limit, Instant.now().plus(expiryDuration));
+                out.printf("Applying throttled rate limit %f expires at %s:\n", limit, throttle.getExpirationTime());
```

Comment:
Is there a reason to use `stdout` directly rather than log4j?

Reply:
That's not `System.stdout`, it's an output stream that's passed as a parameter to the task.  What's written to that stream is the response to the caller.

Reply:
Whoops I missed that